sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala

```diff
@@ -418,115 +443,140 @@ case class SortMergeJoinExec(
     // Inline mutable state since not many join operations in a task
     val matches = ctx.addMutableState(clsName, "matches",
       v => s"$v = new $clsName($inMemoryThreshold, $spillThreshold);", forceInline = true)
-    // Copy the left keys as class members so they could be used in next function call.
-    val matchedKeyVars = copyKeys(ctx, leftKeyVars)
+    // Copy the streamed keys as class members so they could be used in next function call.
+    val matchedKeyVars = copyKeys(ctx, streamedKeyVars)
+
+    // Handle the case when streamed rows has any NULL keys.
+    val handleStreamedAnyNull = joinType match {
+      case _: InnerLike =>
+        // Skip streamed row.
+        s"""
+           |$streamedRow = null;
+           |continue;
+         """.stripMargin
+      case LeftOuter | RightOuter =>
+        // Eagerly return streamed row.
+        s"""
+           |if (!$matches.isEmpty()) {
+           |  $matches.clear();
+           |}
+           |return false;
+         """.stripMargin
+      case x =>
+        throw new IllegalArgumentException(
+          s"SortMergeJoin.genScanner should not take $x as the JoinType")
+    }
 
-    ctx.addNewFunction("findNextInnerJoinRows",
+    // Handle the case when streamed keys less than buffered keys.
+    val handleStreamedLessThanBuffered = joinType match {
+      case _: InnerLike =>
+        // Skip streamed row.
+        s"$streamedRow = null;"
+      case LeftOuter | RightOuter =>
+        // Eagerly return with streamed row.
+        "return false;"
+      case x =>
+        throw new IllegalArgumentException(
+          s"SortMergeJoin.genScanner should not take $x as the JoinType")
+    }
+
+    ctx.addNewFunction("findNextJoinRows",
       s"""
-         |private boolean findNextInnerJoinRows(
-         |    scala.collection.Iterator leftIter,
-         |    scala.collection.Iterator rightIter) {
-         |  $leftRow = null;
+         |private boolean findNextJoinRows(
```

Comment:
It looks reusing the inner-case code makes the outer-case code inefficient. For example, if there are too many matched duplicate rows in the buffered side, it seems we don't need to put all the rows in `matches`, right?

Reply:
> In the outer case, a return value is not used?

Yes. Otherwise it's very hard to re-use code in `findNextJoinRows`. I can further make more change to not return anything for `findNextJoinRows` in case it's an outer join. Do we want to do that?

Reply:
> For example, if there are too many matched duplicate rows in the buffered side, it seems we don't need to put all the rows in matches, right?

Why we don't need to put all the rows? We anyway need to evaluate all the rows on buffered side for join, right?

Reply:
> Why we don't need to put all the rows? We anyway need to evaluate all the rows on buffered side for join, right?

Oh, my bad. ya, you're right. I misunderstood it.

Reply:
> In the outer case, a return value is not used?
Yes. Otherwise it's very hard to re-use code in findNextJoinRows. I can further make more change to not return anything for findNextJoinRows in case it's an outer join. Do we want to do that?

okay, the current one looks fine. Let's just wait for a @cloud-fan comment here.

Reply:
btw, in the current generated code, it seems `conditionCheck` is evaluated outside `findNextJoinRows`. We cannot evaluate it inside `findNextJoinRows` to avoid actual putting unmached rows in `matches`?

Reply:
@maropu - No I think we need buffer anyway. The buffered rows has same join keys with current streamed row. But there can be multiple followed streamed rows having same join keys, as the buffered rows. Even though buffered rows cannot match condition with current streamed rows, they may match condition with followed streamed rows. I think this is how current sort merge join (code-gen & iterator) is designed.